.github/workflows/ci.yml

```diff
@@ -0,0 +1,118 @@
+    strategy:
+      matrix:
+        python-version: ["3.8", "3.10"]
+
+
+
+    steps:
+    - uses: actions/checkout@v2
+    - name: Set up Python ${{ matrix.python-version }}
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ matrix.python-version }}
+    - name: Install Python dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install coverage
+        cd ./sdk
+        pip install .
+    - name: Test with coverage + unittest
+      run: |
+        cd ./sdk
+        coverage run -m unittest
+    - name: Report test coverage
+      if: ${{ always() }}
+      run: |
+        cd ./sdk
+        coverage report -m
+
+  static-analysis:
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v2
+    - name: Set up Python ${{ env.X_PYTHON_VERSION }}
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ env.X_PYTHON_VERSION }}
+    - name: Install Python dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install pycodestyle mypy
+        cd ./sdk
+        pip install .
+    - name: Check typing with MyPy
+      run: |
+        mypy sdk --exclude sdk/build
+    - name: Check code style with PyCodestyle
+      run: |
+        pycodestyle --count --max-line-length 120 sdk
+
+  readme-codeblocks:
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v2
+    - name: Set up Python ${{ env.X_PYTHON_VERSION }}
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ env.X_PYTHON_VERSION }}
+    - name: Install Python dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install pycodestyle mypy codeblocks
+        cd ./sdk
+        pip install .
+    - name: Check typing with MyPy
+      run: |
+        mypy <(codeblocks python README.md)
+    - name: Check code style with PyCodestyle
+      run: |
+        codeblocks --wrap python README.md | pycodestyle --count --max-line-length 120 -
+    - name: Run readme codeblocks with Python
+      run: |
+        codeblocks python README.md | python
+
+  docs:
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v2
+    - name: Set up Python ${{ env.X_PYTHON_VERSION }}
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ env.X_PYTHON_VERSION }}
+    - name: Install Python dependencies
+      run: |
+        python -m pip install --upgrade pip
+      # TODO add autodoc test
+
+
+  package:
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v2
+    - name: Set up Python ${{ env.X_PYTHON_VERSION }}
+      uses: actions/setup-python@v2
+      with:
+        python-version: ${{ env.X_PYTHON_VERSION }}
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip
+        pip install -q build
+    - name: Create source and wheel dist
+      run: |
+        ls
+        cd ./sdk
+        python -m build
```

Comment:
There seems to be a missing empty line at the end of the file

Reply:
Do .yml files have style conventions? If so we could also include a codestyle test for that.

Reply:
Actually the new line at the end of files is not programming style, but stems from the [POSIX definition of a line](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206). Therefore this rule applies to all text files. Apparently, some older unix tools and some git diff tools have problems if that new line character is missing. 

(I love learning the actual reason behind things I've just blindly accepted my whole life :D)  

Of course, yaml files have also a specified syntax, but imo it'd be overkill to check this for configuration files that either work or don't.